scorecard/profile_data/api_data.py

```diff
@@ -178,17 +178,33 @@ def muni_contact(self):
 
         return values
 
-    def audit_opinions(self):
+    def audit_opinions(self, last_opinion_year):
+        years = range(last_opinion_year-3, last_opinion_year+1)
+        audit_results = self.results["audit_opinions"]
         values = []
-        for result in self.results["audit_opinions"]:
-            values.append(
-                {
-                    "date": result["financial_year_end.year"],
-                    "result": result["opinion.label"],
-                    "rating": result["opinion.code"],
-                    "report_url": result["opinion.report_url"],
-                }
-            )
+
+        for year in years:
+            for i, dict in enumerate(audit_results):
+                if dict["financial_year_end.year"] == year:
+                    values.append(
+                        {
+                            "date": audit_results[i]["financial_year_end.year"],
+                            "result": audit_results[i]["opinion.label"],
+                            "rating": audit_results[i]["opinion.code"],
+                            "report_url": audit_results[i]["opinion.report_url"],
+                        }
+                    )
+            if values:
+                if values[-1]["date"] != year:
+                    values.append(
+                        {
+                            "date": year,
+                            "result": None,
+                            "rating": None,
+                            "report_url": None,
+                        }
+                    )
```

Comment:
this is pretty clunky. This is saying "if we find the year, then add it to the list of years. and if the last year we added isn't the current year, then add a placeholder year"

I think what would make more sense is

1. loop over the data you have. for each item, add it to a dict keyed on year
2. loop over the years you expect.
  - if you have it in your dict, add it to your output list
  - if you don't, add a placeholder to your output list

Reply:
What made this messy was that audit opinions were stored as a list of dictionaries. Changed that to a dictionary keyed by year

This means a profile compilation is needed for these changes to work